fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/FieldKeyExpression.java

```diff
@@ -315,6 +315,29 @@ public SplitKeyExpression split(int splitSize) {
         return field.getMessageType();
     }
 
+    /**
+     * Only supporting DOUBLE, INT64, INT32, and BOOL currently.
+     *
+     * @param descriptor Record Descriptor
+     * @return boolean
+     */
+    @Override
+    public boolean isFixedWidth(Descriptors.Descriptor descriptor) {
+        Descriptors.FieldDescriptor fieldDescriptor = descriptor.findFieldByName(fieldName);
+        if (fieldDescriptor.isRepeated() || fieldDescriptor.isOptional()) {
```

Comment:
`isOptional` is probably too strong, as it is often used for things that are, in fact, `NOT NULL`, due to the exact behavior of Protobuf. (Reasonable advice is "always use `optional`.") A better check is probably using the `nullStandin`, which will need to be `NULL_UNIQUE` for nulls to work in a primary key, I think.

Reply:
I think you need `nullStandIn` to be `NOT_NULL` for this to work here, unless there's a representation of `null` that we can think of that can be encoded in the same number of bytes as the non-`null` field values (or can exclude the nullable field somehow)

Reply:
Ok, I tried Mike's approach but I will let you guys discuss the definition of a null field in the system that can be used in a primary key.

Reply:
I think the salient thing for this code is that the requirement is that the field is not null in order for field to be encoded using the fixed point encoder. I think using `NULL_UNIQUE` will continue to create a primary key with a `null` field in the tuple, which is why I don't think it helps us here. In theory, this could become a runtime check, maybe, though we probably want to try harder to not include the grouping keys in the Lucene-encoded primary key (if they overlap), in that case. (See: `primaryKeyComponentPositions` on the index. I think something like only only including the primary key components where the `primaryKeyComponentPosition < 0 || >= groupingCount` would suffice.)

Reply:
My thought was that only `NULL_UNIQUE` really works as a component in a unique index or primary key. While it is true that you could have a `null` there with `NULL`, for example, there would have to be something else that arranged for ignoring that to _still_ be unique. Such as the other columns by themselves being unique in that case. I don't see any particular downside to requiring that primary key fields that might hold nulls must have that (and adding a runtime check as you suggest). Then we could interpret field not so annotated as `NOT NULL`. Lots of databases don't allow nullable fields in primary keys at all, so having an extension to do that seems okay (to me).

Having said all that, I am happy to take a more conservative approach for now and revisit this when nullability is overhauled as part of a more robust type system.